staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache_test.go

```diff
@@ -0,0 +1,216 @@
+	assert := assert.New(t)
+	// Check basic structure variables are set.
+	assert.Equal(testCacheDir, cache.cacheDir)
+	assert.Equal(testCacheSize, cache.cacheSize)
+	// Check "Get()" for non-existent key
+	_, found := cache.Get("non-existent")
+	assert.False(found, "Non-existent file should not be found in cache")
+	// Set file1 into the disk cache, and validate we can "Get()" it.
+	cache.Set("file1", file1)
+	actualFile1, found := cache.Get("file1")
+	assert.True(found, "Expected 'file1' to be found in cache")
+	assert.Equal(file1, actualFile1)
+	// Set file2 into the disk cache, and validate we can "Get()" it.
+	cache.Set("file2", file2)
+	actualFile2, found := cache.Get("file2")
+	assert.True(found, "Expected 'file2' to be found in cache")
+	assert.Equal(file2, actualFile2)
+	// Set file3 into the disk cache, and validate we can "Get()" it
+	// and the other files.
+	cache.Set("file3", file3)
+	_, found = cache.Get("file1")
+	assert.True(found, "Expected 'file1' to be found in cache")
+	_, found = cache.Get("file2")
+	assert.True(found, "Expected 'file2' to be found in cache")
+	actualFile3, found := cache.Get("file3")
+	assert.True(found, "Expected 'file3' to be found in cache")
+	assert.Equal(file3, actualFile3)
+	// Delete file2, and validate it is no longer in the cache, while
+	// file1 and file3 are still in the cache.
+	cache.Delete("file2")
+	_, found = cache.Get("file2")
+	assert.False(found, "Expected 'file2' to NOT be found in cache")
+	_, found = cache.Get("file1")
+	assert.True(found, "Expected 'file1' to be found in cache")
+	_, found = cache.Get("file3")
+	assert.True(found, "Expected 'file3' to be found in cache")
+}
+
+func TestDiskCacheCompaction(t *testing.T) {
+	testCacheDir, err := ioutil.TempDir("", "compaction-cache")
+	defer os.RemoveAll(testCacheDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+	var file1 = []byte("abcdefghijklmno") // 15 chars
+	var file2 = []byte("1234567890")      // 10 chars
+	var file3 = []byte("fghijklm")        // 8 chars
+	var file4 = []byte("0987654321")      // 10 chars
+	assert := assert.New(t)
+	fullFilePath1 := filepath.Join(testCacheDir, keyToFilename("file1"))
+	err = ioutil.WriteFile(fullFilePath1, file1, 0644)
+	assert.NoError(err)
+	fullFilePath2 := filepath.Join(testCacheDir, keyToFilename("file2"))
+	err = ioutil.WriteFile(fullFilePath2, file2, 0644)
+	assert.NoError(err)
+	fullFilePath3 := filepath.Join(testCacheDir, keyToFilename("file3"))
+	err = ioutil.WriteFile(fullFilePath3, file3, 0644)
+	assert.NoError(err)
+	fullFilePath4 := filepath.Join(testCacheDir, keyToFilename("file4"))
+	err = ioutil.WriteFile(fullFilePath4, file4, 0644)
+	assert.NoError(err)
+	// Files by oldest modTime: file2, file4, file1, file3
+	now := time.Now()
+	setFileModTime(t, fullFilePath1, now.Add(5*time.Hour))
+	setFileModTime(t, fullFilePath2, now.Add(-10*time.Hour))
+	setFileModTime(t, fullFilePath3, now.Add(10*time.Hour))
+	setFileModTime(t, fullFilePath4, now.Add(-5*time.Hour))
```

Comment:
Personally I would test this by hiding the filesystem functions behind an interface, so one could test the logic without actually needing to put stuff on disk (via a fake implementation). Tests that hit disk just tend to be slow/flaky.

Up to you.

Reply:
Would it be possible to refactor these tests (and other discovery client tests) with an in-memory file system in a subsequent PR?

Reply:
personally, I would do it now or not at all...